subprojects/architecture-test/src/test/java/org/gradle/architecture/test/PublicApiAccessTest.java

```diff
@@ -0,0 +1,187 @@
+import com.tngtech.archunit.core.domain.JavaType;
+import com.tngtech.archunit.core.domain.JavaTypeVariable;
+import com.tngtech.archunit.core.domain.JavaWildcardType;
+import com.tngtech.archunit.junit.AnalyzeClasses;
+import com.tngtech.archunit.junit.ArchTest;
+import com.tngtech.archunit.lang.ArchCondition;
+import com.tngtech.archunit.lang.ArchRule;
+import com.tngtech.archunit.lang.ConditionEvents;
+import com.tngtech.archunit.lang.SimpleConditionEvent;
+import groovy.lang.Closure;
+import groovy.util.Node;
+import groovy.xml.MarkupBuilder;
+import org.w3c.dom.Element;
+
+import javax.xml.namespace.QName;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.Reader;
+import java.io.StringWriter;
+import java.io.Writer;
+import java.math.BigDecimal;
+import java.net.URI;
+import java.net.URL;
+import java.time.Duration;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+
+import static com.tngtech.archunit.core.domain.JavaClass.Predicates.resideInAnyPackage;
+import static com.tngtech.archunit.core.domain.JavaClass.Predicates.type;
+import static com.tngtech.archunit.core.domain.JavaMember.Predicates.declaredIn;
+import static com.tngtech.archunit.core.domain.JavaModifier.PUBLIC;
+import static com.tngtech.archunit.core.domain.properties.HasModifiers.Predicates.modifier;
+import static com.tngtech.archunit.lang.conditions.ArchConditions.not;
+import static com.tngtech.archunit.lang.conditions.ArchPredicates.are;
+import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.classes;
+import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.methods;
+import static org.gradle.architecture.test.ArchUnitFixture.freeze;
+import static org.gradle.architecture.test.ArchUnitFixture.gradleInternalApi;
+import static org.gradle.architecture.test.ArchUnitFixture.gradlePublicApi;
+import static org.gradle.architecture.test.ArchUnitFixture.haveDirectSuperclassOrInterfaceThatAre;
+
+@AnalyzeClasses(packages = "org.gradle")
+public class PublicApiAccessTest {
+
+    private static final DescribedPredicate<JavaMember> public_api_methods = declaredIn(gradlePublicApi())
+        .and(modifier(PUBLIC))
+        .as("public API methods");
+
+    private static final DescribedPredicate<JavaClass> primitive = new DescribedPredicate<JavaClass>("primitive") {
+        @Override
+        public boolean apply(JavaClass input) {
+            return input.isPrimitive();
+        }
+    };
+
+    private static final DescribedPredicate<JavaClass> allowed_types_for_public_api =
+        gradlePublicApi()
+            .or(primitive)
+            .or(resideInAnyPackage("java.lang", "java.util", "java.util.concurrent", "java.util.regex", "java.util.function", "java.lang.reflect", "java.io")
+                .or(type(File[].class))
+                .or(type(FileInputStream.class))
```

Comment:
Isn't `type(FileInputStream.class)` duplicate of `resideInAnyPackage("java.io")`?

Reply:
You are right! I removed the ones already covered by `java.io`.